lib/ansible/module_utils/azure_rm_common.py

```diff
@@ -0,0 +1,635 @@
+        :return: None
+        '''
+        # tags must be a dictionary of string:string mappings.
+        if not isinstance(tags, dict):
+            self.fail("Tags must be a dictionary of string:string values.")
+        for key, value in tags.iteritems():
+            if not isinstance(value, str):
+                self.fail("Tags values must be strings. Found {0}:{1}".format(str(key), str(value)))
+
+    def _tag_purge(self, tags):
+        '''
+        Remove metadata tags not found in user provided tags parameter. Returns tuple
+        with bool indicating something changed and dict of new tags to be assigned to
+        the object.
+
+        :param tags: object metadata tags
+        :return: bool, dict of tags
+        '''
+        if not self.module.params.get('tags'):
+            # purge all tags
+            return True, dict()
+        new_tags = copy.copy(tags)
+        changed = False
+        for key in tags:
+            if not self.module.params['tags'].get(key):
+                # key not found in user provided parameters
+                new_tags.pop(key)
+                changed = True
+        if changed:
+            self.log('CHANGED: purged tags')
+        return changed, new_tags
+
+    def _tag_update(self, tags):
+        '''
+        Update metadata tags with values in user provided tags parameter. Returns
+        tuple with bool indicating something changed and dict of new tags to be
+        assigned to the object.
+
+        :param tags: object metadata tags
+        :return: bool, dict of tags
+        '''
+        if isinstance(tags, dict):
+            new_tags = copy.copy(tags)
+        else:
+            new_tags = dict()
+        changed = False
+        if self.module.params.get('tags'):
+            for key, value in self.module.params['tags'].iteritems():
+                if not (new_tags.get(key) and new_tags[key] == value):
+                    changed = True
+                    new_tags[key] = value
+        if changed:
+            self.log('CHANGED: updated tags')
+        return changed, new_tags
+
+    def update_tags(self, tags):
+        '''
+        Call from the module to update metadata tags. Returns tuple
+        with bool indicating if there was a change and dict of new
+        tags to assign to the object.
+
+        :param tags: metadata tags from the object
+        :return: bool, dict
+        '''
+        changed = False
+        updated, new_tags = self._tag_update(tags)
+        if updated:
+            changed = True
+
+        if self.module.params['purge_tags']:
+            purged, new_tags = self._tag_purge(new_tags)
+            if purged:
+                changed = True
+        return changed, new_tags
+
+    def exec_module(self):
+        '''
+        Execute module implementation.
+
+        :return: None
+        '''
+        res = self.exec_module_impl(**self.module.params)
+        self.module.exit_json(**res)

```

Comment:
A better pattern for this is to make exec_module an abstract method (ie: one that the subclasses must implement).  Then have the subclass call self.module.exit_json() directly.

In python-2.6+ you can use the abc library: https://docs.python.org/2/library/abc.html or for something quick you can just do:
``` python
def exec_module(self):
    raise NotImplementedError()
```

The subclasses then implement exec_module and do **not** call the base class's exec_module method.